File: testerum-backend/testerum-runner/testerum-runner-api/src/main/kotlin/com/testerum/runner/report_model/ReportFeature.kt
package com.testerum.runner.report_model

import com.fasterxml.jackson.annotation.JsonProperty
import com.testerum.api.test_context.ExecutionStatus
import java.time.LocalDateTime

data class ReportFeature(@JsonProperty("featureName")      val featureName: String,
                         @JsonProperty("tags")             val tags: List<String>,
                         @JsonProperty("startTime")        val startTime: LocalDateTime,
                         @JsonProperty("endTime")          val endTime: LocalDateTime,
                         @JsonProperty("durationMillis")   val durationMillis: Long,
                         @JsonProperty("status")           val status: ExecutionStatus,
                         @JsonProperty("textLogFilePath")  override val textLogFilePath: String,
                         @JsonProperty("modelLogFilePath") override val modelLogFilePath: String,
                         @JsonProperty("children")         val children: List<FeatureOrTestRunnerReportNode>) : FeatureOrTestRunnerReportNode

File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/report-model/model/report/report-feature.ts
import {FeatureOrTestRunnerReportNode} from "./feature-or-test-runner-report-node";
import {ExecutionStatus} from "./execution-status";
import {MarshallingUtils} from "../../../json-marshalling/marshalling-utils";
import {RunnerReportNodeType} from "./runner-report-node";
import {ReportTest} from "./report-test";
import {ReportParametrizedTest} from "./report-parametrized-test";
import {ReportHooks} from "./report-hooks";

export class ReportFeature implements FeatureOrTestRunnerReportNode {

    constructor(public readonly featureName: string,
                public readonly tags: Array<string>,
                public readonly startTime: Date,
                public readonly endTime: Date,
                public readonly durationMillis: number,
                public readonly status: ExecutionStatus,
                public readonly textLogFilePath: string,
                public readonly modelLogFilePath: string,
                public readonly children: Array<FeatureOrTestRunnerReportNode>) {
    }

    static parse(input: Object): ReportFeature {
        if (!input) {
            return null;
        }

        const featureName = input["featureName"];
        const tags = MarshallingUtils.parseListOfStrings(input["tags"]);
        const startTime = MarshallingUtils.parseUtcToLocalDateTime(input["startTime"]);
        const endTime = MarshallingUtils.parseUtcToLocalDateTime(input["endTime"]);
        const durationMillis = input["durationMillis"];
        const status = MarshallingUtils.parseEnum(input["status"], ExecutionStatus);
        const textLogFilePath = input["textLogFilePath"];
        const modelLogFilePath = input["modelLogFilePath"];
        const children = MarshallingUtils.parseListPolymorphically<FeatureOrTestRunnerReportNode>(input["children"], {
            [RunnerReportNodeType[RunnerReportNodeType.FEATURE]]: ReportFeature,
            [RunnerReportNodeType[RunnerReportNodeType.TEST]]: ReportTest,
            [RunnerReportNodeType[RunnerReportNodeType.PARAMETRIZED_TEST]]: ReportParametrizedTest,
            [RunnerReportNodeType[RunnerReportNodeType.REPORT_HOOKS]]: ReportHooks
        });

        return new ReportFeature(featureName, tags, startTime, endTime, durationMillis, status, textLogFilePath, modelLogFilePath, children);
    }

}
